### PR TITLE
SAAS-3834: Added AWS Config Referencing Missing S3 Bucket Plugin

### PR DESCRIPTION
**Pull request by sadeed12345**

### CODE REVIEW

File: plugins/aws/configservice/referencingMissingS3Bucket.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+module.exports = {
+    title: 'Config Service Enabled',
+    category: 'ConfigService',
+    domain: 'Management and Governance',
+    description: 'Ensure that Amazon Config service is pointing an S3 bucket that is active in your account in order to save configuration information',
+    more_info: 'Amazon Config tracks changes within the configuration of your AWS resources and it regularly sends updated configuration details to an S3 bucket that you specify.'+
+        'When AWS Config is not referencing an active S3 bucket, the service is unable to send the recorded information to the designated bucket, therefore you lose the ability to audit later the configuration changes made within your AWS account.',
+    recommended_action: 'Enable that Amazon Config service is referencing an active S3 bucket in order to save configuration information.',
+    link: 'https://aws.amazon.com/config/details/',
```
Comment (sadeed12345) [inline plugins/aws/configservice/referencingMissingS3Bucket.js:12]:
```suggestion
    link: 'https://docs.aws.amazon.com/config/latest/developerguide/s3-bucket-policy.html',
```

File: plugins/aws/configservice/referencingMissingS3Bucket.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    domain: 'Management and Governance',
+    description: 'Ensure that Amazon Config service is pointing an S3 bucket that is active in your account in order to save configuration information',
+    more_info: 'Amazon Config tracks changes within the configuration of your AWS resources and it regularly sends updated configuration details to an S3 bucket that you specify.'+
+        'When AWS Config is not referencing an active S3 bucket, the service is unable to send the recorded information to the designated bucket, therefore you lose the ability to audit later the configuration changes made within your AWS account.',
+    recommended_action: 'Enable that Amazon Config service is referencing an active S3 bucket in order to save configuration information.',
+    link: 'https://docs.aws.amazon.com/config/latest/developerguide/s3-bucket-policy.html',
+    apis: ['ConfigService:describeDeliveryChannels', 'S3:headBucket'],
+  
+    run: function(cache, settings, callback) {
+        // console.log(JSON.stringify(cache, null, 2));
```
Comment (sadeed12345) [inline plugins/aws/configservice/referencingMissingS3Bucket.js:16]:
```suggestion
```

File: plugins/aws/configservice/referencingMissingS3Bucket.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            for (let record of describeDeliveryChannels.data) {
+
+                var headBucket = helpers.addSource(cache, source,
+                    ['s3', 'headBucket', region, record.s3BucketName]);
+
+                if (!headBucket.data || !headBucket) {
+                    helpers.addResult(results, 0,
+                        'the referenced S3 bucket is available within your AWS account',
+                        region);
+
```
Comment (m-nouman-suleman) [inline plugins/aws/configservice/referencingMissingS3Bucket.js:46]:
```suggestion
```

File: plugins/aws/configservice/referencingMissingS3Bucket.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+                helpers.addResult(results, 0, 'No Config delivery channels found', region);
+                return rcb();
+            }
+
+            for (let record of describeDeliveryChannels.data) {
+
+                var headBucket = helpers.addSource(cache, source,
+                    ['s3', 'headBucket', region, record.s3BucketName]);
+
+                if (!headBucket.data || !headBucket) {
```
Comment (m-nouman-suleman) [inline plugins/aws/configservice/referencingMissingS3Bucket.js:42]: Is the data from headBucket coming in the .data object?

File: plugins/aws/configservice/referencingMissingS3Bucket.spec.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+const referencingMissingS3Bucket = require('./referencingMissingS3Bucket');
+
+const describeDeliveryChannels = [
+    {
+        "name": "default",
+        "s3BucketName": "config-bucket-000011112222",
+        "s3KeyPrefix": "mine"
+    }
+];
+
```
Comment (m-nouman-suleman) [inline plugins/aws/configservice/referencingMissingS3Bucket.spec.js:11]:
```suggestion
```

File: plugins/aws/configservice/referencingMissingS3Bucket.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+                } else if (!headBucket || headBucket.err) {
+                    helpers.addResult(results, 3,
+                        'Unable to query S3 headbucket: ' + helpers.addError(headBucket), region);
+                    continue;
+                }
+            }
+
+            if (deletedBucket.length) {
+                helpers.addResult(results, 2,
+                    `config service have these buckets  ${deletedBucket.join(', ')}, not available in your account`,
```
Comment (sadeed12345) [inline plugins/aws/configservice/referencingMissingS3Bucket.js:57]:
```suggestion
                    `Config Service have these buckets  ${deletedBucket.join(', ')}, not available in your account`,
```

File: plugins/aws/configservice/referencingMissingS3Bucket.js [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+                    region);
+
+            } else {
+                helpers.addResult(results, 0,
+                    'config service have buckets available in your account',
```
Comment (sadeed12345) [inline plugins/aws/configservice/referencingMissingS3Bucket.js:62]:
```suggestion
                    'Config Service have buckets available in your account', 
```

File: plugins/aws/configservice/referencingMissingS3Bucket.spec.js [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+            referencingMissingS3Bucket.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].region).to.equal('us-east-1');
+                expect(results[0].message).to.include('config service have buckets available in your account')
+                done();
+            });
+        });
+
+        it('should FAIL if config service have these buckets not available in your account.', function (done) {
```
Comment (sadeed12345) [inline plugins/aws/configservice/referencingMissingS3Bucket.spec.js:55]:
```suggestion
        it('should FAIL if Config Service have these buckets not available in your account.', function (done) {
```

File: plugins/aws/configservice/referencingMissingS3Bucket.spec.js [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+                }
+            }
+
+        },
+    }
+}
+
+describe('referencingMissingS3Bucket', function () {
+    describe('run', function () {
+        it('should PASS if config service have these buckets available in your account.', function (done) {
```
Comment (sadeed12345) [inline plugins/aws/configservice/referencingMissingS3Bucket.spec.js:44]:
```suggestion
        it('should PASS if Config Service have these buckets available in your account.', function (done) {
```

File: plugins/aws/configservice/referencingMissingS3Bucket.js [hunk 10]
```diff
@@ -0,0 +1,70 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Config Service Enabled',
+    category: 'ConfigService',
+    domain: 'Management and Governance',
+    description: 'Ensure that Amazon Config service is pointing an S3 bucket that is active in your account in order to save configuration information',
+    more_info: 'Amazon Config tracks changes within the configuration of your AWS resources and it regularly sends updated configuration details to an S3 bucket that you specify.'+
```
Comment (sadeed12345) [inline plugins/aws/configservice/referencingMissingS3Bucket.js:9]:
```suggestion
    more_info: 'Amazon Config tracks changes within the configuration of your AWS resources and it regularly sends updated configuration details to an S3 bucket that you specify. '+
```

**Review by AkhtarAmir — APPROVED** [review]

**Pull request merged**